rpki-validator/src/main/java/net/ripe/rpki/validator3/domain/validation/CertificateTreeValidationService.java

```diff
@@ -291,69 +290,108 @@ private boolean isValidationRunCompleted(ValidationResult validationResult) {
 
             temporary.setLocation(new ValidationLocation(manifestUri));
-            manifest.validate(manifestUri.toASCIIString(), context, crl.get(), manifest.getCrlUri(), VALIDATION_OPTIONS, temporary);
+            manifest.validate(manifestUri.toASCIIString(), context, x509Crl, manifest.getCrlUri(), VALIDATION_OPTIONS, temporary);
             if (temporary.hasFailureForCurrentLocation()) {
                 return validatedObjects;
             }
             validatedObjects.add(manifestObject.get().key());
 
-            Bench.mark(trustAnchor.getName(), "retrieveManifestEntries", () ->
-                storage.readTx(tx -> retrieveManifestEntries(tx, manifest, manifestUri, temporary)))
-                .entrySet()
-                .stream().map(e -> {
-                URI location = e.getKey();
-                RpkiObject rpkiObject = e.getValue();
-                temporary.setLocation(new ValidationLocation(location));
-
-                final Optional<CertificateRepositoryObject> maybeCertificateRepositoryObject = Bench.mark(trustAnchor.getName(),
-                    "rpkiObject.get", () -> rpkiObject.get(CertificateRepositoryObject.class, temporary));
-
-                if (!temporary.hasFailureForCurrentLocation()) {
-                    return maybeCertificateRepositoryObject.flatMap(certificateRepositoryObject -> {
-                        Bench.mark0(trustAnchor.getName(), "certificateRepositoryObject.validate", () ->
-                            certificateRepositoryObject.validate(location.toASCIIString(), context, crl.get(), crlUri, VALIDATION_OPTIONS, temporary));
-
-                        if (!temporary.hasFailureForCurrentLocation()) {
-                            validatedObjects.add(rpkiObject.key());
-                        }
-
-                        if (certificateRepositoryObject instanceof X509ResourceCertificate
-                            && ((X509ResourceCertificate) certificateRepositoryObject).isCa()
-                            && !temporary.hasFailureForCurrentLocation()) {
-
-                            return Optional.of(context.createChildContext(location, (X509ResourceCertificate) certificateRepositoryObject));
-                        }
-                        return Optional.empty();
-                    });
-                }
-                return Optional.<CertificateRepositoryObjectValidationContext>empty();
-            })
-                .filter(Optional::isPresent)
-                .map(Optional::get)
-                .collect(Collectors.toList())
+            manifest.getFiles().entrySet()
                 .parallelStream()
-                .map(childContext -> validateCertificateAuthority(trustAnchor, registeredRepositories, childContext, temporary))
-                .forEachOrdered(validatedObjects::addAll);
+                .map(entry -> getManifestEntry(manifestUri, entry))
+                .filter(t -> t.v2().isPresent())
+                .map(t -> t.map2(Optional::get))
```

Comment:
Instead of map filter optional, I think flatmap and returning stream/empty stream is nicer, see commit on separate branch:
https://github.com/RIPE-NCC/rpki-validator-3/commit/9dd3321d80c21e2a13f4941d976a0807e388fe72

If it make sense you can merge it here, not worth separate PR.